#' Convert sequence files between various formats
#' @param infile A file of type genbank, embl (not tested), fasta, fastq, ecopcrdb.
#' @param in_type Accepted values: "genbank", "embl", "fasta", "fastq", "ecopcrdb".
#' @param taxo An obitools-formatted taxonomy database as an R object, generated by \code{\link[ROBITaxonomy]{read.taxonomy}}
#'     or \code{\link[bastools]{NCBI2obitaxonomy}}. This is only used when creating an ecopcrdb - taxids must be in header in format
#'     > seqX taxids=8276
#'     see \code{obiaddtaxids.Bas}.
#' @param skip_errors logical. skips entries where taxonomy could not be found. Only used when creating an ecopcrdb.
#' ########################Hangs without this...make mandatory. also taxids are mandatory
#' @param out_type Accepted values: "--fasta-output", "--ecopcrdb-output"
#' @note Creating an ecopcrdb takes much longer here than running in a terminal, not sure why! (e.g. 20 seqs took).
#'     Looks like an issue creating the .adx file of new taxonomy.
#' @return A file of type fasta or ecopcrdb
#' @examples
#' fastq->fasta
#' a<-system.file("extdata", "head.test.minion.fastq", package = "bastools")
#' obitaxonomydb<-"/media/sf_Documents/WORK/CIBIO/STATS_AND_CODE/TAXONOMIES/obitax_26-4-19"
#' obiconvert.Bas2(a,taxo = obitaxonomydb, out_type = "--fasta-output",out="head.test.minion.fasta",in_type="fastq")
#' ok
#'
#' fastq->ecopcrdb, error
#' ok
#'
#' fasta without taxids->ecopcrdb
#' a<-system.file("extdata", "head.test.fasta", package = "bastools")
#' obitaxonomydb<-"/media/sf_Documents/WORK/CIBIO/STATS_AND_CODE/TAXONOMIES/obitax_26-4-19"
#' obiconvert.Bas(a,taxo = obitaxonomydb, out_type = "--ecopcrdb-output",out="head.test.ecopcrdb",in_type="fasta")
#'#############hangs
#'
#' fasta with taxids->ecopcrdb
#' a<-system.file("extdata", "wTaxids.head20.test.fasta", package = "bastools")
#' obitaxonomydb<-"/media/sf_Documents/WORK/CIBIO/STATS_AND_CODE/TAXONOMIES/obitax_26-4-19"
#' obiconvert.Bas(a,taxo = obitaxonomydb, out_type = "--ecopcrdb-output",out="wTaxids.head.test.ecopcrdb",in_type="fasta")
#' #################NOT WORKING PROPERLY, SUPER SLOW.....
#' #14:27 start (20 seqs with taxids)
#'
#' genbank->fasta
#' a<-system.file("extdata", "head_ALL_VERTS_REFSEQ_MTDNA.gb", package = "bastools")
#' obitaxonomydb<-"/media/sf_Documents/WORK/CIBIO/STATS_AND_CODE/TAXONOMIES/obitax_26-4-19"
#' obiconvert.Bas(a,taxo = obitaxonomydb, out_type = "--ecopcrdb-output",out="head.test.ecopcrdb")
#'
#' genbank to ecopcrdb
#'
#' @export
obiconvert.Bas<-function(infile,in_type,taxo,out_type,out){
  if(in_type!="fasta" & out_type=="--ecopcrdb-output") stop("Only fasta files can be converted into an ecopcrdb")

  cb <- function(line, proc) {cat(line, "\n")}
  if(out_type=="--fasta-output"){
    b<-processx::run(command = "obiconvert", args=c(infile,"-d",taxo,out_type),
                     echo=F,stderr_line_callback = cb,echo_cmd = T)
    writeLines(b$stdout,con = out)
  }
  if(out_type=="--ecopcrdb-output"){
    message("Make sure input fasta file has taxids in sequence headers in the format \" taxid=XXXX;\"")
    processx::run(command = "obiconvert", args=c(infile,"-d",taxo,paste0(out_type,"=",out),"--skip-on-error"),
                  echo=F,stderr_line_callback = cb,echo_cmd = T)
    d<-"SUCCESS!"
  }
  h<-processx::run(command = "obicount", args=c(infile),echo=F,echo_cmd = F,stderr_line_callback = cb)
  j<-stringr::str_locate_all(string = h$stdout,pattern = " " )[[1]][,1]
  k<-substr(h$stdout,1,j-1)
  l<-paste0("input_read_count=",k)
  h<-processx::run(command = "obicount", args=c(out),echo=F,echo_cmd = F,stderr_line_callback = cb)
  j<-stringr::str_locate_all(string = h$stdout,pattern = " " )[[1]][,1]
  k<-substr(h$stdout,1,j-1)
  m<-paste0("output_read_count=",k)
  print("SUCCESS!")
  print(l)
  print(m)
}

#' @export
bascount.fasta<-function(infile){
  f<-processx::run(command = "grep",args = c("-c",'>',infile),echo_cmd = T,echo = T)
  as.numeric(f$stdout)
}

#' @export
bascount.fastas.recur<-function(folder){
  cb <- function(line, proc) {cat(line, "\n")}
  origpath<-getwd()
  setwd(folder)
  a<-list.files(pattern = "*.fasta",recursive = T)
  b<-processx::run(command = "grep",args = c("-cH",'>',a),echo_cmd = F,echo = F,stderr_line_callback = cb)
  d<-read.table(text = b$stdout,sep = "\n")
  print(d)
  e<-stringr::str_split(string = d$V1,pattern = ":")
  setwd(origpath)
  sum(sapply(e, function(x){as.numeric(x[2])}))

}


