---
title: "Comparing 2 taxatables"
html_notebook: default
subtitle: ''
classoption: portrait
output: html_document
---

## read and print settings
```{r read and print settings, eval=TRUE,echo=F}

```


```{r load functions, eval=TRUE, results='hide',echo=F}
source("/home/tutorial/TOOLS/bastools/master_functions.R")
library(ggplot2)
options(httr_oob_default = TRUE)
all.stats<-list()
stepcounter<-0
```

## import taxatabs
```{r import taxatabs, eval=TRUE,echo=F, results='hide'}
#import taxtabs
taxatab1<-bas.merge.taxatabs(c("/media/sf_Documents/WORK/CIBIO/AA_PROJECTS/MUSSELS/March_2020/2019_August_002.UNIO.lenFilt.trimmed.ids.SC3.taxatab.txt"
                             ,"/media/sf_Documents/WORK/CIBIO/AA_PROJECTS/MUSSELS/March_2020/2019_August_002.VENE.lenFilt.trimmed.ids.SC3.taxatab.txt"
                             ,"/media/sf_Documents/WORK/CIBIO/AA_PROJECTS/MUSSELS/March_2020/2019_September_001.UNIO.lenFilt.trimmed.ids.SC3.taxatab.txt"
                             ,"/media/sf_Documents/WORK/CIBIO/AA_PROJECTS/MUSSELS/March_2020/2019_September_001.VENE.lenFilt.trimmed.ids.SC3.taxatab.txt"))

name.taxatab1<-"Raw; 7 target species"

taxatab2<-bas.merge.taxatabs(c("/media/sf_Documents/WORK/CIBIO/AA_PROJECTS/MUSSELS/March_2020/2019_August_002.UNIO.lenFilt.trimmed.ids.SC3.pol.taxatab.txt"
                             ,"/media/sf_Documents/WORK/CIBIO/AA_PROJECTS/MUSSELS/March_2020/2019_August_002.VENE.lenFilt.trimmed.ids.SC3.pol.taxatab.txt"
                             ,"/media/sf_Documents/WORK/CIBIO/AA_PROJECTS/MUSSELS/March_2020/2019_September_001.UNIO.lenFilt.trimmed.ids.SC3.pol.taxatab.txt"
                             ,"/media/sf_Documents/WORK/CIBIO/AA_PROJECTS/MUSSELS/March_2020/2019_September_001.VENE.lenFilt.trimmed.ids.SC3.pol.taxatab.txt"))

name.taxatab2<-"Polished; 7 target species"

#import taxtabs
taxatab3<-bas.merge.taxatabs(c("/media/sf_Documents/WORK/CIBIO/AA_PROJECTS/MUSSELS/March_2020/2019_August_002.UNIO.lenFilt.trimmed.ids.SC1.taxatab.txt"
                             ,"/media/sf_Documents/WORK/CIBIO/AA_PROJECTS/MUSSELS/March_2020/2019_August_002.VENE.lenFilt.trimmed.ids.SC1.taxatab.txt"
                             ,"/media/sf_Documents/WORK/CIBIO/AA_PROJECTS/MUSSELS/March_2020/2019_September_001.UNIO.lenFilt.trimmed.ids.SC1.taxatab.txt"
                             ,"/media/sf_Documents/WORK/CIBIO/AA_PROJECTS/MUSSELS/March_2020/2019_September_001.VENE.lenFilt.trimmed.ids.SC1.taxatab.txt"))

name.taxatab3<-"Raw; 16S Bivalves"

taxatab4<-bas.merge.taxatabs(c("/media/sf_Documents/WORK/CIBIO/AA_PROJECTS/MUSSELS/March_2020/2019_August_002.UNIO.lenFilt.trimmed.ids.SC1.pol.taxatab.txt"
                             ,"/media/sf_Documents/WORK/CIBIO/AA_PROJECTS/MUSSELS/March_2020/2019_August_002.VENE.lenFilt.trimmed.ids.SC1.pol.taxatab.txt"
                             ,"/media/sf_Documents/WORK/CIBIO/AA_PROJECTS/MUSSELS/March_2020/2019_September_001.UNIO.lenFilt.trimmed.ids.SC1.pol.taxatab.txt"
                             ,"/media/sf_Documents/WORK/CIBIO/AA_PROJECTS/MUSSELS/March_2020/2019_September_001.VENE.lenFilt.trimmed.ids.SC1.pol.taxatab.txt"))

name.taxatab4<-"Polished; 16S Bivalves"

#Plotting taxonomic levels attained
taxatab.list<-list(taxatab1,taxatab2,taxatab3,taxatab4)
names(taxatab.list)<-c(name.taxatab1,name.taxatab2,name.taxatab3,name.taxatab4)
taxlevel_prop_out<-"/media/sf_Documents/WORK/CIBIO/AA_PROJECTS/MUSSELS/March_2020/prop_levels.png"

#OPTIONAL - COMPARE BLAST HITS, just change compare.blast.hits=F to ignore
compare.blast.hits=F

blastfile1<-"/media/sf_Documents/WORK/CIBIO/AA_PROJECTS/MUSSELS/March_2020/2019_August_002.VENE.lenFilt.trimmed.ids.SC1.pol.blast.txt"
blastfile2<-"/media/sf_Documents/WORK/CIBIO/AA_PROJECTS/MUSSELS/March_2020/2019_August_002.VENE.lenFilt.trimmed.ids.SC3.pol.blast.txt"

name.blastfile1<-"Aug.VENE.SC1.pol"
name.blastfile2<-"Aug.VENE.SC3.pol"

cutoff=99

pident_col="V5"
qseqid_col="V1"

```

## reads / sample
```{r reads per sample, eval=TRUE,echo=F}
#general - no. samples, no. taxa
#1. reads / sample
sample.reads<-compare.sample.reads(taxatab1,taxatab2,rm.nohits = T,rm.NAs = T)
print(sample.reads[[1]])
print(sample.reads[[2]])
print(sample.reads[[3]])

```

## number of reads attaining species level, genus level, family level
```{r number of reads attaining level, eval=TRUE,echo=F}
#plotting function takes a named list and can be any number of taxa tables
a<-plot.taxatab.rank.props(taxatab.list)
print(a)
ggsave(filename = taxlevel_prop_out,plot = a,device = "png")

#stats function can onyl take 2 taxatables, as data frames 
tax.levels<-chisq.taxatab.ranks(taxatab1,taxatab2,name.taxatab1,name.taxatab2)
print(tax.levels[[1]])
print(tax.levels[[2]])
```

## species accumulation curves 
```{r species accumulation curve, eval=TRUE,echo=F}
print(bas.plot.specaccum(taxatab = taxatab1))
print(bas.plot.specaccum(taxatab = taxatab2))

print(bas.plot.specrich(taxatab = taxatab1))
print(bas.plot.specrich(taxatab = taxatab2))
```

## comapare detections
```{r compare detections, eval=TRUE,echo=F}
#ideas
#highlight differences in binary
binarytaxatab1<-binarise.taxatab(taxatab1,t = T)
binarytaxatab2<-binarise.taxatab(taxatab2,t=T)

compd1d2 <- dataCompareR::rCompare(binarytaxatab1, binarytaxatab2)
summary(compd1d2)
dataCompareR::generateMismatchData(compd1d2,binarytaxatab1,binarytaxatab2)

#highlight differences in taxa in each sample...number of identical detections...number of non-identical detections
binarytaxatab1<-binarise.taxatab(taxatab1)
binarytaxatab2<-binarise.taxatab(taxatab2)

binarytaxatab1<-keep.below.xLevel.assigns(binarytaxatab1)

binarytaxatab2<-keep.below.xLevel.assigns(binarytaxatab2)

for(i in 2:length(colnames(binarytaxatab1[,-1]))){
  samplex<-binarytaxatab1[,c(1,i)]
  message(colnames(samplex)[2])
  samplexdxn<-samplex[samplex[,2]==1,]
  
  if(is.na(match(colnames(samplex)[2],colnames(binarytaxatab2)))) {
    message("sample not in 2nd taxatable. taxa detected in 1st taxatable")
    print(samplexdxn$taxon)
    
  } else {
    sampley<-binarytaxatab2[,c(1,match(colnames(samplexdxn)[2],colnames(binarytaxatab2)))]
    sampleydxn<-sampley[sampley[,2]==1,]
    
    message("taxa detected in ",colnames(samplexdxn)[2]," in table1, not in table 2:")
    print(samplexdxn$taxon[!samplexdxn$taxon %in% sampleydxn$taxon])
    
    message("taxa detected in ",colnames(sampleydxn)[2]," in table2, not in table 1:")
    print(sampleydxn$taxon[!sampleydxn$taxon %in% samplexdxn$taxon])
  }
}
    
    
  sample1<-binarytaxatab1[,c(1,i)]
binarytaxatab1[binarytaxatab1$taxon=="Eukaryota;Mollusca;Bivalvia;Unionida;Unionidae;Anodonta;Anodonta exulcerata",]



```

## compare blast hits 
```{r compare blast hits, eval=TRUE,echo=F}
if(compare.blast.hits) {
  blast.comparison<-blast.maxhits.2.files.compare(blastfile1,blastfile2,name.blastfile1,name.blastfile2,pident_col=pident_col,qseqid_col=qseqid_col,cutoff)
  print(blast.comparison[[1]])
  print(blast.comparison[[2]])
  print(blast.comparison[[3]])
}
```